File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys
input = sys.stdin.readline

while True:
    s = input().rstrip()
    if s == '.': break

    stack = []
    done = 1
    for i in range(len(s)):
        if s[i] == '(': stack.append(1)
        elif s[i] == '[': stack.append(2)
        elif s[i] == ')': 
            if len(stack) == 0 or stack[-1] == 2:
                done = 0
                break
            else: stack.pop()
        elif s[i] == ']':
            if len(stack) == 0 or stack[-1] == 1:
                done = 0
                break
            else: stack.pop()
        
    if len(stack) != 0: done = 0
    print('yes' if done == 1 else 'no')

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.c
#include <stdio.h>
#include <string.h>

int main(void) {
    while(1) {
        char str[105];
        gets(str);
        if(!strcmp(str, ".")) break;
        int stack[105] = {0,};
        int stackSize = 0;
        int done = 1;
        for (int i = 0; i < strlen(str); i++) {
            if (str[i] == '(') stack[stackSize++] = 1;
            else if (str[i] == ')') {
                if (stackSize == 0 || stack[stackSize - 1] == 2) {
                    done = 0;
                    break;
                }
                else stackSize -= 1;
            }
            else if (str[i] == '[') stack[stackSize++] = 2;
            else if (str[i] == ']') {
                if (stackSize == 0 || stack[stackSize - 1] == 1) {
                    done = 0;
                    break;
                }
                else stackSize -= 1;
            }
        }

        if (stackSize > 0) done = 0;

        if (done) printf("yes\n");
        else printf("no\n");
    }
    return 0;
}